File: app/src/main/java/com/worldventures/dreamtrips/wallet/service/firmware/command/LoadAppAtmelFirmwareCommand.java
package com.worldventures.dreamtrips.wallet.service.firmware.command;

import io.techery.janet.command.annotations.CommandAction;
import io.techery.janet.smartcard.action.support.UpgradeIntAtmelFirmwareAction;
import io.techery.janet.smartcard.event.UpgradeIntAtmelFirmwareProgressEvent;
import rx.Observable;

import com.worldventures.dreamtrips.wallet.domain.entity.ImmutableSmartCardFirmware;
import com.worldventures.dreamtrips.wallet.domain.entity.SmartCardFirmware;

import java.io.File;

@CommandAction
public class LoadAppAtmelFirmwareCommand extends BaseLoadFirmwareCommand {

   private final File firmwareFile;
   private final String firmwareVersion;

   public LoadAppAtmelFirmwareCommand(File firmwareFile, String firmwareVersion) {
      this.firmwareFile = firmwareFile;
      this.firmwareVersion = firmwareVersion;
   }

   @Override
   Observable<Integer> provideProgress() {
      return janet.createPipe(UpgradeIntAtmelFirmwareProgressEvent.class)
            .observeSuccess()
            .map(event -> event.progress);
   }

   @Override
   Observable<Void> loadFile() {
      return janet.createPipe(UpgradeIntAtmelFirmwareAction.class)
            .createObservableResult(new UpgradeIntAtmelFirmwareAction(firmwareFile))
            .map(action -> null);
   }

   @Override
   SmartCardFirmware updatedSmartCardFirmware(SmartCardFirmware currentSmartCardFirmware) {
      return ImmutableSmartCardFirmware.copyOf(currentSmartCardFirmware).withInternalAtmelVersion(firmwareVersion);
   }

}

File: wallet/src/main/java/com/worldventures/wallet/service/firmware/command/LoadAppAtmelFirmwareCommand.kt
package com.worldventures.wallet.service.firmware.command

import com.worldventures.wallet.domain.entity.SmartCardFirmware
import io.techery.janet.command.annotations.CommandAction
import io.techery.janet.smartcard.action.support.UpgradeIntAtmelFirmwareAction
import io.techery.janet.smartcard.event.UpgradeIntAtmelFirmwareProgressEvent
import rx.Observable
import java.io.File

@CommandAction
class LoadAppAtmelFirmwareCommand(private val firmwareFile: File, private val firmwareVersion: String) : BaseLoadFirmwareCommand() {

   override fun provideProgress(): Observable<Int> {
      return janet.createPipe(UpgradeIntAtmelFirmwareProgressEvent::class.java)
            .observeSuccess()
            .map { event -> event.progress }
   }

   override fun loadFile(): Observable<Void> {
      return janet.createPipe(UpgradeIntAtmelFirmwareAction::class.java)
            .createObservableResult(UpgradeIntAtmelFirmwareAction(firmwareFile))
            .map { null }
   }

   override fun updatedSmartCardFirmware(currentSmartCardFirmware: SmartCardFirmware): SmartCardFirmware {
      return currentSmartCardFirmware.copy(internalAtmelVersion = firmwareVersion)
   }

}
